### PR TITLE
fix(lsp): analyze fs dependencies of dependencies to find npm package requirements

### LINKED ISSUE #16867: @deno-types doesn't work until focusing the file that contains the comment
**Describe the bug**

I'm re-exporting an npm module in `mithril.ts`; notice the `@deno-types` comment:

```ts
// @deno-types='npm:@types/mithril@2.0.11'
export { default } from 'npm:mithril@2.2.2'
```

I have the file mapped to `mithril` in `import_map.json` (and I have `{ "importMap": "./import_map.json" }` in `deno.json`):

```json
{
  "imports": {
    "mithril": "./mithril.ts"
  }
}
```

...so that in `app.ts` I can import [Mithril.js](https://mithril.js.org/) like this:

```ts
import m from 'mithril'

m.mount()
```

The `m` variable has the correct type (via `@deno-types`) – the Deno extension correctly complains that `m.mount()` expects two parameters.

And here's the bug:

If I reload (close and reopen) VS Code, the `m` variable in `app.ts` has type `any` until I open and focus `mithril.ts`. After focusing `mithril.ts`, the `m` variable in `app.ts` has the correct type, even if I close `mithril.ts` and keep it closed.

**To Reproduce**

1. Add those two TS files to your project and configure the import mapping
2. Focus any other file than `mithril.ts`. `mithril.ts` can be left open as long as its tab is not focused. Alternatively close all files
3. Reload VS Code e.g. via Command Palette: "Developer: Reload Window"
4. Open/focus `app.ts` and notice how `m` has type `any`
5. Open/focus `mithril.ts`
6. Open/focus `app.ts` again and notice how `m` has the correct type

**Expected behavior**

I would expect the 5th step above to be unnecessary – the `m` variable in `app.ts` should have the correct type already in step 4, without the need to focus `mithril.ts`.

**Screenshots**

N/A

**Versions**

- VS Code: 1.73.1
- Deno: 1.28.1
- VS Code extension: 3.14.1

```bash
$ deno --version
deno 1.28.1 (release, aarch64-apple-darwin)
v8 10.9.194.1
typescript 4.8.3
```

### PR DESCRIPTION
**Pull request by dsherret**
Closes #16867

### CODE REVIEW

File: cli/lsp/documents.rs [hunk 1]
```diff
@@ -633,6 +634,23 @@ struct FileSystemDocuments {
...
+    let fs_version = get_document_path(cache, specifier)
+      .and_then(|path| calculate_fs_version(&path));
+    let file_system_doc = self.docs.get(specifier);
+    if file_system_doc.map(|d| d.fs_version().to_string()) != fs_version {
+      // attempt to update the file on the file system
+      self.refresh_document(cache, get_maybe_resolver(), specifier)
+    } else {
+      file_system_doc.cloned()
+    }
+  }
```
Comment (dsherret) [inline cli/lsp/documents.rs:652]: Extracted out from below.

File: cli/lsp/documents.rs [hunk 2]
```diff
@@ -633,6 +634,23 @@ struct FileSystemDocuments {
 }
 
 impl FileSystemDocuments {
+  pub fn get<'a>(
+    &mut self,
+    cache: &HttpCache,
+    get_maybe_resolver: impl Fn() -> Option<&'a dyn deno_graph::source::Resolver>,
```
Comment (bartlomieju) [inline cli/lsp/documents.rs:640]: Does it actually need to be a closure? It seems it's used only in this function so should be okay to pass actual resolver?

**Review by bartlomieju — APPROVED** [review]
LGTM, nice fix

File: cli/lsp/documents.rs (hunk 2 above)
Comment (dsherret) [reply cli/lsp/documents.rs:640]: I looked into it and it seems ok to just create it each time. It's essentially just providing a reference. I'll change this.

**Pull request merged**